Library/Homebrew/cmd/update-bash.sh

```diff
@@ -2,22 +2,38 @@ brew() {
   "$HOMEBREW_BREW_FILE" "$@"
 }
 
+git() {
+  "$HOMEBREW_GIT" "$@"
+}
+
 which_git() {
-  local which_git
   local active_developer_dir
 
-  which_git="$(which git 2>/dev/null)"
-  if [[ -n "$which_git" && "/usr/bin/git" = "$which_git" ]]
+  if [[ -n "$HOMEBREW_GIT" && -x "$HOMEBREW_GIT" ]]
+  then
+    echo "$HOMEBREW_GIT"
+    return
+  fi
+
+  if [[ -n "$GIT" && -x "$GIT" ]]
+  then
+    HOMEBREW_GIT="$GIT"
+    echo "$HOMEBREW_GIT"
+    return
+  fi
+
+  HOMEBREW_GIT="$(which git 2>/dev/null)"
```

Comment:
Given the task this function is now performing I think it should be transformed into the shell equivalent of `Utils.ensure_git_installed!` and invoked at the beginning of a shell command that requires `git`. And the newly introduced `git` shell function itself might be best off failing hard if this setup function was never executed.

> I think it would be nice to maybe have some of this as a set environment variable in `bin/brew`.

:+1: If we can do that in a way that doesn't slow down commands that don't require `git` at all.